<#56> feat : 사전답안 API 구현

nrudev opened the pull request:
## 내용
사전답안 CRUD 기능 구현
- 작성
- 기술면접 및 사용자별 조회
- 수정
- 삭제

## 참고
- #56

nrudev committed:
feat : PreAnswer 저장 기능 구현

nrudev committed:
feat : PreAnswer 사전답안 및 사용자별 조회 기능 구현

nrudev committed:
feat : PreAnswer 수정 기능 구현

nrudev committed:
feat : PreAnswer 삭제 기능 구현

nrudev committed:
Merge branch 'develop' of https://github.com/J2KB-5th/chippo into feature/56-back-preAnswerApi

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/PreAnswerController.java:
```diff
@@ -22,51 +26,69 @@
+        Long userId = user == null ? 0L : user.getUserId();
+
+        PreAnswerResponse response = null;
+        try {
+            response = new PreAnswerResponse(preAnswerService.getOnePreAnswer(interviewId, userId));
+        } catch (NullPointerException e) {}
```
혹시 제가 남긴 커밋 보셨을까요...?
PR은 아니지만 interviewFindAPI 브랜치에 일반 커밋으로 올려두었는데, 예외 클래스를 생성해서 해당 메시지를 받아올 수 있게끔 구현 했어요. 이건 병합 후 제가 일괄 리팩토링 할게요! 다만 커밋 내용 확인해주시면 좋을 것 같네요.

nrudev replied:
아 이건 에러 메시지를 전달해줄 게 아니라 예외 발생 없이 비어있는 상태 그대로 있는 게 맞아서 이렇게 처리했어요.
이 부분 회의때 말씀드릴게요!

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/PreAnswerController.java:
```diff
@@ -22,51 +26,69 @@
+
+        PreAnswerResponse response = null;
+        try {
+            response = new PreAnswerResponse(preAnswerService.getOnePreAnswer(interviewId, userId));
```
또한 service단의 메서드에서 get 네이밍은 잘 사용하지 않는 걸로 알고 있어요! 
컨트롤러와 레포지토리 모두 find 계열의 네이밍을 사용하니 `findPreAnswer` / `findPreAnswers`(혹은 `PreAnswerList`) 정도로 함께 통일하면 어떨까요?

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/PreAnswerController.java:
```diff
@@ -22,51 +26,69 @@
-                LocalDateTime.now()
-        );
-        return ResponseEntity.ok(testPreAnswer);
+        Long userId = user == null ? 0L : user.getUserId();
```
null일 경우 0L을 사용하시는 이유가 궁금합니다. *몰라서 여쭤봅니다...
그리고 스웨거에서 지금 세션유저도 parameter로 인식하여 required request parameter로 처리되는데, 이 부분도 수정이 필요할 것 같아요. 제가 찾아보겠지만 까먹을 것 같아서;; 혹시 다음 PR 때 누락되어있으면 말씀 한 번만 부탁드릴게요.

nrudev replied:
이 부분 저도 정말 많이 고민하고 작성했어요ㅠ
사전답안을 작성하지 않았을 경우 혹은 로그인을 하지 않았을 경우에는 사전답안이 조회되지 않아야 해서 저렇게 처리했습니다.
userId 는 1부터 시작하는 auto_increment 로 설정되어 있기 때문에 0이 나올 수 없거든요.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/PreAnswerController.java:
```diff
@@ -22,51 +26,69 @@
-        URI uri = uriBuilder.path("/api/interviews/{interviewId}/pre-answers").buildAndExpand(interviewId).toUri();
-        PreAnswerResponse response = new PreAnswerResponse(interviewId, preAnswerRequest.getContent(), new UserResponse(10L, "리액트개발자"), LocalDateTime.now());
+        if (user == null || user.getUserId() != preAnswerRequest.getUserId()) {
+            return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
```
user가 null인 경우에는 unAuthorized이지만 유저 정보가 일치하지 않거나, 권한이 없는 경우는 forbidden 응답코드가 정확하다고 생각합니다.  이 케이스만 나눠서 처리해주시면 좋을 것 같고, 유저 처리 정말 잘 해주셨네요!
** 또 몰라서 여쭤봅니다만 role 검사는 안 해도 될까요? BLOCKED 여부는 나중에 고려하면 될까요?

nrudev replied:
BLOCKED 에 대한 개념이 애매해서 role 검사는 우선 제외했습니다.
만약 BLOCKED 일 경우, 로그인을 못하게 막는 거라면 여기서는 유효성 검사가 필요 없을 것 같고, 그게 아니라 게시물 작성 등의 기능만 못하게 할 거라면 추가하겠습니다. 이거는 회의때 논의해보시죠ㅎㅎ

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/PreAnswerController.java:
```diff
@@ -22,51 +26,69 @@
+        @LoginUser SessionUser user
     ){
-        PreAnswerResponse response = new PreAnswerResponse(interviewId, preAnswerRequest.getContent(), new UserResponse(11L, "노드개발자"), LocalDateTime.now());
+        if (user == null || preAnswerRequest.getUserId() != user.getUserId() || preAnswerRequest.getId() != preAnswerId) {
+            return new ResponseEntity<>(HttpStatus.UNAUTHORIZED);
+        }
```
위에 언급한 내용 참고 해주세요 :)

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/PreAnswerController.java:
```diff
@@ -22,51 +26,69 @@
+        @Parameter(description = "사전답안 ID") @PathVariable(name = "preAnswerId") Long preAnswerId,
+        @LoginUser SessionUser user
     ){
+        preAnswerService.deletePreAnswer(interviewId, preAnswerId, user);
```
delete는 권한 검사도 하기 때문에 서비스로 아예 빼신 거죠? 흠... controller로 빼자니 role 검사 때문에 컨트롤러가 무거워지고, 서비스로 빼자니 update, save는 userId만으로 유저 정보를 주고 받는 거에 비해 delete는 객체 전부가 전송되어 불필요한 정보까지 오고가게 되네요. 개인적으로는 이왕 컨트롤러에 넣은 김에 유저 유효성은 컨트롤러에서 전부 처리함이 어떨까 싶습니다. 제 추측이 틀렸거나 의견 있으시면 편하게 말씀 해주세요!

nrudev replied:
user 유효성 검사를 서비스 단에서 하는 게 맞는 거 같습니다.
role이 admin인지 확인하는 것과 삭제를 요청한 사용자와 해당 요청 대상이 되는 사전답안의 작성자 id 가 같은지 여부를 동시에 검사해야 하거든요..

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/repository/PreAnswerRepository.java:
```diff
@@ -2,6 +2,14 @@
+
+    @Query(nativeQuery = true,
+            value = "SELECT * FROM pre_answer WHERE user_id = :user_id AND interview_id = :interview_id")
+    Optional<PreAnswer> findByIdAndUserId(@Param("interview_id") Long interviewId, @Param("user_id") Long userId);
```
`findByInterviewIdAndUserId` 혹시 이걸로 안 되던가요?

nrudev replied:
네이밍 쿼리만을 해결할 수는 있지만 findByIdAndUser 로 해야 하며, 그럴 경우, Long 타입의 userId가 아닌, User 객체 자체를 파라미터로 전달해주어야 해서 저렇게 처리했습니다.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/service/PreAnswerServiceImpl.java:
```diff
@@ -1,12 +1,69 @@
+        return preAnswerRepository.save(request.toEntity(user, interview));
+    }
+
+    @Override
```
`@Transactional(readOnly = true)` 넣어주시면 좋을 것 같아요.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/service/PreAnswerServiceImpl.java:
```diff
@@ -1,12 +1,69 @@
+    @Override
+    public PreAnswer updatePreAnswer(UpdatePreAnswerRequest request, Long interviewId) {
+
+        return preAnswerRepository.findById(request.getId())
+                .orElseThrow(() -> new ResponseStatusException(HttpStatus.BAD_REQUEST, "존재하지 않는 사전답변입니다."))
```
개인적으로 이 부분은 중복되는 코드이니 클래스 내 private 메서드로 추출해도 좋을 것 같습니다. !

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/service/PreAnswerServiceImpl.java:
```diff
@@ -1,12 +1,69 @@
+    }
+
+    @Override
+    public PreAnswer getOnePreAnswer(Long interviewId, Long userId) {
+
+        return preAnswerRepository.findByIdAndUserId(interviewId, userId)
+                .orElseGet(null);
+    }
```
조회 부분 로직이 잘 이해가 안 가네요... 이따 회의 때 시간 괜찮으시면 설명 부탁드리겠습니다 !

nrudev replied:
사전답안이 조회되지 않는 경우가 존재합니다. 예를 들면, 사용자가 사전답안을 작성하지 않은 시점이거나, 사용자가 로그인되어 있지 않은 경우에요.

아! 그냥 .get() 하나 .orElseGet(null) 하나 똑같군요!!
이거 때문에 컨트롤러 단에서 try-catch 로 NullPointerException을 받아줬던 겁니다!
조회 데이터 없어도 예외 발생 안하고, 빈 상태라는 걸 보여주기 위해서요.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/preanswer/controller/PreAnswerController.java:
```diff
@@ -22,51 +26,69 @@
-    @GetMapping("/{preAnswerId}")
+    @Operation(summary = "사전답안 기술면접 및 사용자별 조회",
+            description = "특정 기술면접에 대해 특정 사용자가 등록한 사전답안 게시글이 있다면 함께 전달됩니다.")
+    @GetMapping({"", "/"})
```
몰라서 여쭤봅니다만, `@GetMapping`과 `@GetMapping({"", "/"})`은 어떤 차이가 있나요?

nrudev replied:
큰 차이 없는데 `@GetMapping` 만 적는 게 나을 것 같네요. 수정하겠습니다!

dadahee reviewed (commented):
네이밍, 코드 전부 신경 많이 쓰신 게 눈에 보이네요. 정말 고생 많으셨습니다!

nrudev committed:
refactor : PR#63 리뷰에 따른 코드 수정

- 중복 코드 메서드화
- deletePreAnswer에 user 로그인 여부 확인 추가 및 service로 user 대신 userId 전달
- 조회 메서드 네이밍 변경

nrudev committed:
chore : SessionUser Swagger에서 숨김처리
